### DISCUSSION TITLE
How to prevent instrumentation running during build

### DISCUSSION THREAD

**Discussion by captain-derp** [vercel/next.js#56328] [App Router]
I'm using the instrumentation.ts to initialize infrastructure services such as message queue and database client. This required fetching passwords from secure secret storage such as Azure KeyVault. However the problem is that instrumentation code is executed during build time inside a docker build environment. This means that the code will not have access to the credentials required to retrieve the passwords from KeyVault.

Why does next.js need to run the instrumentation code?
Is there a way to prevent next.js from running the instrumentation code while building?

Here is my Dockerfile:
```
FROM node:18-alpine AS base

# Install dependencies only when needed
FROM base AS deps
RUN apk add --no-cache libc6-compat
WORKDIR /app

COPY package.json yarn.lock* package-lock.json* pnpm-lock.yaml* ./
RUN \
  if [ -f yarn.lock ]; then yarn --frozen-lockfile; \
  elif [ -f package-lock.json ]; then npm ci; \
  elif [ -f pnpm-lock.yaml ]; then yarn global add pnpm && pnpm i --frozen-lockfile; \
  else echo "Lockfile not found." && exit 1; \
  fi

FROM base AS builder
WORKDIR /app
COPY --from=deps /app/node_modules ./node_modules
COPY . .

# When running build, next.js executes instrumentation.ts  
RUN yarn build

FROM base AS runner
WORKDIR /app

ENV NODE_ENV production

RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs

COPY --from=builder /app/public ./public

# Set the correct permission for prerender cache
RUN mkdir .next
RUN chown nextjs:nodejs .next

COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static

USER nextjs

EXPOSE 80

ENV PORT 80
ENV HOSTNAME "0.0.0.0"

CMD ["node", "server.js"]
```

**Comment by oyatek** [comment]
same problem for me

**Comment by truediogo** [comment]
same here, using instrumentation to validate environment variables at server start and getting error on build, because server variables is not set yet..

**Comment by Gregory-Ledray** [comment]
I successfully used https://nextjs.org/docs/app/building-your-application/optimizing/instrumentation#importing-runtime-specific-code

```
export async function register() {
    if (process.env.NEXT_RUNTIME === 'nodejs') {
        // This does NOT run during build
    }
}
```

**Reply by smoothdvd** [reply]
not working on nextjs 15

**Comment by bytekai** [comment]
experiencing same problem

**Comment by IanGraingerGMSL** [comment]
Hating how we can't write code sensibly using env vars in Next.js because they aren't defined at build time. Of course they aren't!

**Comment by yakubka** [comment]
the NEXT_RUNTIME check posted above is the right fix. here is the full picture so the why is clear.

during next build, Next.js evaluates instrumentation.ts once as part of module graph analysis. at that point no server has started yet, so NEXT_RUNTIME is undefined. any code that reaches out to external services at the top of register() will fail because it runs in a build context with no network access to your infrastructure.

NEXT_RUNTIME has three possible values:
- undefined during next build
- nodejs at server runtime for Node.js routes and server components
- edge at runtime for Edge route handlers

the pattern that prevents build-time execution:

```ts
export async function register() {
  if (process.env.NEXT_RUNTIME === "nodejs") {
    // only runs when the server starts, never during next build
    const { initDb } = await import("./lib/db");
    await initDb();
  }
}
```

the dynamic import matters too. if your initialization module imports things that only exist in Node.js (native modules, Azure SDK, etc.) a static top-level import would still be evaluated at build time even inside the if block. the dynamic import defers the require until runtime.

for the environment variable validation case specifically:

```ts
export async function register() {
  if (process.env.NEXT_RUNTIME === "nodejs") {
    const required = ["DATABASE_URL", "AZURE_VAULT_URL", "QUEUE_CONNECTION_STRING"];
    for (const k of required) {
      if (!process.env[k]) throw new Error(`missing required env: ${k}`);
    }
    const { init } = await import("./lib/services");
    await init();
  }
}
```

this pattern runs once per server cold start and never touches the build container.
